botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1027 @@
+    is_valid_ipv4_endpoint_url,
+    is_valid_ipv6_endpoint_url,
+    normalize_url_path,
+    percent_encode,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class BaseRule:
+    """A base class representing a rule within a rule set. All rules contain
+    a conditions property, which can be empty, a set of assigned variables
+    passed by a caller, and an endpoint provider instance.
+    """
+
+    TEMPLATE_STRING_RE = re.compile(r'(?<=\{)[a-zA-Z#]+(?=\})')
+    VALID_HOST_LABEL_RE = re.compile(
+        r"^((?!-)[A-Z\d-]{1,63}(?<!-)\.)*((?!-)[A-Z\d-]{1,63}(?<!-))$",
+        re.IGNORECASE,
+    )
+    GET_ATTR_RE = re.compile(r'(\w+)\[(\d+)\]')
+    FUNCTION_NAME_MAPPING = {
+        'aws.partition': 'partition',
+        'aws.parseArn': 'parse_arn',
+        'not': '_not',
+    }
+
+    def __init__(self, conditions, assignments, provider):
+
+        self.conditions = conditions
+        self.assignments = assignments
+        self.provider = provider
+
+    def evaluate(self):
+        raise NotImplementedError()
+
+    def is_func(self, argument):
+        """Determine if an object is a function object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a function object
+
+        :returns: A bool indicating if the input is a function object
+        """
+        return isinstance(argument, dict) and 'fn' in argument
+
+    def is_ref(self, argument):
+        """Determine if an object is a reference object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a reference object
+
+        :returns: A bool indicating if the input is a reference object
+        """
+        return isinstance(argument, dict) and 'ref' in argument
+
+    def is_template(self, argument):
+        """Determine if an object contains a template string.
+
+        :type argument: Any
+        :param argument: An object that may or may not contain a template string.
+
+        :returns: A bool indicating if the input contains a template string.
+        """
+        return (
+            isinstance(argument, str)
+            and self.TEMPLATE_STRING_RE.search(argument) is not None
+        )
+
+    def resolve_template_string(self, value):
+        """Given a templated string, parse out the template nested between
+        curly braces, and return the nested data from the rule's assignments.
+
+        :type value: str
+        :param value: A template string e.g. "https://{Bucket}.ec2.{url#authority}"
+
+        :returns: A fully resolved string e.g. "https://us-west-2.ec2.amazonaws.com"
+        """
+        matches = self.TEMPLATE_STRING_RE.findall(value)
+        for match in matches:
+            template_params = match.split('#')
+            template_value = self.assignments
+            for param in template_params:
+                template_value = template_value[param]
+            # f strings containing curly braces do not get parsed correctly
+            # by string.replace
```

Comment:
ok yeah unfortunately double curly braces breaks an f string. it treats it like the string literal `"{match}"` so it looks for that instead of the value and replaces nothing

Reply:
i changed the regex to include the curly braces and strip them off later so this won't be an issue